### DISCUSSION TITLE
Using a Dependency Manager

### DISCUSSION THREAD

**Discussion by AliSayyah** [piccolo-orm/piccolo#586] [Ideas]
Hey there!
I wanted to propose we start using a proper dependency manager for the project.
Apart from, of course, managing the dependencies, it would simply packaging and development cycles too.
I've worked with `poetry` and `PDM`, and both of them are pretty good, but I prefer `poetry` as it seems more stable, and I found out that it has better `GH actions` support.
I would like to hear your thoughts about it!

**Comment by dantownsend** [comment]
It's an interesting idea.

I used Poetry and PipEnv in the past. I remember one of them being really slow, but apparently things have improved recently.

What do you think the main things problems they solve for open source projects? I scanned through the Poetry docs again, and it does seem to have a lot of features.

**Reply by AliSayyah** [reply]
Yeah, the slow one is PipEnv. We had to get rid of it at work because of being too slow.
For example, the need for multiple requierements.txt files would be gone,
and all development commands would be consistent. `poetry run lint`, `poetry run format`, etc. No need to call shell scripts. Easier to remember for me, at least.
Installing development dependencies would be a single command instead of creating a venv and calling `pip install` three times.
Overall, it would give us a cleaner development and setup phase.

**Reply by AliSayyah** [reply]
And everything is handled in `pyproject.toml` and not scattered all over the place. 
These all are, of course, my opinion and can differ from person to person.

**Reply by dantownsend** [reply]
You're right - we do use a lot of shell scripts for managing various things.

Seems like Poetry is a solid option. I need to go through the docs again and familiarise myself with it.

**Reply by AliSayyah** [reply]
Yeah, alright. I can make a quick PR to see how it would go.

**Reply by dantownsend** [reply]
Cool, thanks

**Comment by AliSayyah** [comment]
I was counting on the [grouping ](https://python-poetry.org/docs/master/managing-dependencies/)functionality of poetry. But it only exists in the `development` version and not the stable version.
I know it works, but it may not be a good idea to depend on it. Maybe we should wait for it to become stable.

**Reply by dantownsend** [reply]
Yeah, probably best. We definitely need that feature.
